data/transactions/logic/assembler.go

```diff
@@ -0,0 +1,1112 @@
+		}
+		ops.Out.WriteByte(0x2c)
+		ops.Out.WriteByte(uint8(val))
+	}
+	ops.tpush(StackBytes)
+	return nil
+}
+
+// Txn writes opcodes for loading a field from the current transaction
+func (ops *OpStream) Txn(val uint64) error {
+	if val >= uint64(len(TxnFieldNames)) {
+		return errors.New("invalid txn field")
+	}
+	ops.Out.WriteByte(0x31)
+	ops.Out.WriteByte(uint8(val))
+	ops.tpush(TxnFieldTypes[val])
+	return nil
+}
+
+// Gtxn writes opcodes for loading a field from the current transaction
+func (ops *OpStream) Gtxn(gid, val uint64) error {
+	if val >= uint64(len(TxnFieldNames)) {
+		return errors.New("invalid txn field")
+	}
+	if gid > 255 {
+		return errors.New("gtxn cannot look up beyond group index 255")
+	}
+	ops.Out.WriteByte(0x33)
+	ops.Out.WriteByte(uint8(gid))
+	ops.Out.WriteByte(uint8(val))
+	ops.tpush(TxnFieldTypes[val])
+	return nil
+}
+
+// Global writes opcodes for loading an evaluator-global field
+func (ops *OpStream) Global(val uint64) error {
+	if val >= uint64(len(GlobalFieldNames)) {
+		return errors.New("invalid txn field")
+	}
+	ops.Out.WriteByte(0x32)
+	ops.Out.WriteByte(uint8(val))
+	ops.tpush(GlobalFieldTypes[val])
+	return nil
+}
+
+func assembleInt(ops *OpStream, args []string) error {
+	val, err := strconv.ParseUint(args[0], 0, 64)
+	if err != nil {
+		return err
+	}
+	return ops.Uint(val)
+}
+
+// Explicit invocation of const lookup and push
+func assembleIntC(ops *OpStream, args []string) error {
+	constIndex, err := strconv.ParseUint(args[0], 0, 64)
+	if err != nil {
+		return err
+	}
+	return ops.Intc(uint(constIndex))
+}
+func assembleByteC(ops *OpStream, args []string) error {
+	constIndex, err := strconv.ParseUint(args[0], 0, 64)
+	if err != nil {
+		return err
+	}
+	return ops.Bytec(uint(constIndex))
+}
+
+func parseBinaryArgs(args []string) (val []byte, consumed int, err error) {
+	arg := args[0]
+	if strings.HasPrefix(arg, "base32(") || strings.HasPrefix(arg, "b32(") {
+		open := strings.IndexRune(arg, '(')
+		close := strings.IndexRune(arg, ')')
+		if close == -1 {
+			err = errors.New("byte base32 arg lacks close paren")
+			return
+		}
+		val, err = base32.StdEncoding.WithPadding(base32.NoPadding).DecodeString(arg[open+1 : close])
+		if err != nil {
+			return
+		}
+		consumed = 1
+	} else if strings.HasPrefix(arg, "base64(") || strings.HasPrefix(arg, "b64(") {
+		open := strings.IndexRune(arg, '(')
+		close := strings.IndexRune(arg, ')')
+		if close == -1 {
+			err = errors.New("byte base64 arg lacks close paren")
+			return
+		}
+		val, err = base64.StdEncoding.DecodeString(arg[open+1 : close])
+		if err != nil {
+			return
+		}
+		consumed = 1
+	} else if strings.HasPrefix(arg, "0x") {
+		val, err = hex.DecodeString(arg[2:])
+		if err != nil {
+			return
+		}
+		consumed = 1
+	} else if arg == "base32" || arg == "b32" {
+		if len(args) < 2 {
+			err = fmt.Errorf("need literal after 'byte %s'", arg)
+			return
+		}
+		val, err = base32.StdEncoding.WithPadding(base32.NoPadding).DecodeString(args[1])
```

Comment:
The assembler seems doesn't allow padding for b32 but allow padding for b64 input. Could you elaborate why?

Reply:
our common text representation for addresses is unpadded base32

Reply:
For addresses (`addr`) this makes sense. However, arbitrary base 32-encoded data will be decoded incorrectly (wrong length). Do we want to clarify that we don't accept arbitrary encodings here? (Seems less misleading to me to just drop `base32` in this case.)

Reply:
For addresses (`addr`) this is fine, but for arbitrary `base32` encoded data this will silently fail to decode correctly (wrong length).

Reply:
If we want to avoid confusion, it makes sense to me to either (1) drop `base32` as an encoding format for TEAL assembly or (2) have `base32` behave differently from `addr` and take standard padding.

Reply:
unpadded base32 is clean. for 1 to 11 bytes I encoded to unpadded base32 string and decoded back and got the same bytes.

Reply:
assembler `byte ...` will now accept base32 strings with constant padding